src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/CounterGroup.cs

```diff
@@ -120,58 +122,82 @@ internal static CounterGroup GetCounterGroup(EventSource eventSource)
+        private bool _isPollingEnabled;
 
-        private void DisposeTimer()
-        {
-            Debug.Assert(Monitor.IsEntered(this));
-            if (_pollingTimer != null)
-            {
-                _pollingTimer.Dispose();
-                _pollingTimer = null;
-            }
-        }
-
-        private void EnableTimer(float pollingIntervalInSeconds)
+        private void EnablePollingThread(float pollingIntervalInSeconds)
         {
             Debug.Assert(Monitor.IsEntered(this));
             if (pollingIntervalInSeconds <= 0)
             {
-                DisposeTimer();
                 _pollingIntervalInMilliseconds = 0;
+                _isPollingEnabled = false;
             }
             else if (_pollingIntervalInMilliseconds == 0 || pollingIntervalInSeconds * 1000 < _pollingIntervalInMilliseconds)
             {
-                Debug.WriteLine("Polling interval changed at " + DateTime.UtcNow.ToString("mm.ss.ffffff"));
                 _pollingIntervalInMilliseconds = (int)(pollingIntervalInSeconds * 1000);
-                DisposeTimer();
                 _timeStampSinceCollectionStarted = DateTime.UtcNow;
                 // Don't capture the current ExecutionContext and its AsyncLocals onto the timer causing them to live forever
                 bool restoreFlow = false;
                 try
                 {
+                    if (_pollingEnabledEvent == null)
+                    {
+                        _pollingEnabledEvent = new ManualResetEvent(false);
+                    }
+
                     if (!ExecutionContext.IsFlowSuppressed())
                     {
                         ExecutionContext.SuppressFlow();
                         restoreFlow = true;
                     }
 
-                    _pollingTimer = new Timer(s => ((CounterGroup)s!).OnTimer(null), this, _pollingIntervalInMilliseconds, _pollingIntervalInMilliseconds);
+                    if (_pollingThread == null)
+                    {
+                        Thread t = new Thread(PollForValues) { IsBackground = true };
+                        t.Start();
+                        _pollingThread = t;
+                    }
+                }
+                catch (OutOfMemoryException)
+                {
+                    _isPollingEnabled = false; // If we can't start the polling thread, disable polling.
                 }
                 finally
                 {
                     // Restore the current ExecutionContext
                     if (restoreFlow)
                         ExecutionContext.RestoreFlow();
+
+                    if (_pollingEnabledEvent != null)
+                        _pollingEnabledEvent.Set();
+                }
+            }
+        }
+
+        private void PollForValues()
+        {
+            while (true)
+            {
+                while (_isPollingEnabled)
+                {
+                    if (!OnTimer())
+                    {
+                        break;
```

Comment:
Its not clear to me why we need two methods of detecting that polling should stop. Are you aware of a case where checking _isPollingEnabled alone doesn't do the right thing?